elastic/kibana
# [Fleet] Package policy upgrade telemetry with sender

## Summary

Collecting telemetry of package policy upgrades: https://github.com/elastic/kibana/issues/109870

See latest commits for Event Sender implementation (in this case we don't need collector).
Took it from security_solution and simplified for our use case.

See `sender.ts`
To test:
Install older version of packages, e.g. apache-0.3.3 (no conflicts), aws-0.6.1 (conflicts).
Add policies for each and update them.
Aws policy can only be upgraded manually by resolving conflicts.

For event sender:
Add to `kibana.dev.yml` to opt in telemetry and see debug logs:
```
telemetry.optIn: true
logging.loggers:
 - name: plugins.fleet
   level: debug
```
In case of success/failure of upgrades, the debug log should include events sent to telemetry v3 endpoint.

```
[2021-10-18T16:49:12.391+02:00][DEBUG][plugins.fleet.telemetry_events] Telemetry URL: https://telemetry-staging.elastic.co/v3-dev/send/fleet-stats
[2021-10-18T16:49:12.391+02:00][DEBUG][plugins.fleet.telemetry_events] cluster_uuid: vvUkWqalQJara7GMDQnE3Q cluster_name: elasticsearch
[2021-10-18T16:49:12.392+02:00][DEBUG][plugins.fleet.telemetry_events] [{"package_policy_upgrade":{"package_name":"aws","current_version":"0.6.1","new_version":"1.3.0","status":"failure","error":[{"key":"inputs.cloudtrail-aws-s3.streams.aws.cloudtrail.vars.queue_url","message":["Queue URL is required"]},{"key":"inputs.cloudwatch-aws-s3.streams.aws.cloudwatch_logs.vars.queue_url","message":["Queue URL is required"]},{"key":"inputs.ec2-aws-s3.streams.aws.ec2_logs.vars.queue_url","message":["Queue URL is required"]},{"key":"inputs.elb-aws-s3.streams.aws.elb_logs.vars.queue_url","message":["Queue URL is required"]},{"key":"inputs.s3-aws-s3.streams.aws.s3access.vars.queue_url","message":["Queue URL is required"]},{"key":"inputs.vpcflow-aws-s3.streams.aws.vpcflow.vars.queue_url","message":["Queue URL is required"]},{"key":"inputs.waf-aws-s3.streams.aws.waf.vars.queue_url","message":["Queue URL is required"]}]},"id":"aws_0.6.1_1.3.0_failure","license":{"uid":"c0f3d843-30fd-44f2-9774-95e02db80d70","status":"active","type":"trial","issued_to":"elasticsearch","issuer":"elasticsearch"},"cluster_uuid":"vvUkWqalQJara7GMDQnE3Q","cluster_name":"elasticsearch"}]
[2021-10-18T16:49:12.801+02:00][DEBUG][plugins.fleet.telemetry_events] Events sent!. Response: 200 {"status":"ok"}
```




### Checklist

Delete any items that are not applicable to this PR.



### Risk Matrix

Delete this section if it is not applicable to this PR.

Before closing this PR, invite QA, stakeholders, and other developers to identify risks that should be tested prior to the change/feature release.

When forming the risk matrix, consider some of the following examples and how they may potentially impact the change:

| Risk                      | Probability | Severity | Mitigation/Notes        |
|---------------------------|-------------|----------|-------------------------|
| Multiple Spaces&mdash;unexpected behavior in non-default Kibana Space. | Low | High | Integration tests will verify that all features are still supported in non-default Kibana Space and when user switches between spaces. |
| Multiple nodes&mdash;Elasticsearch polling might have race conditions when multiple Kibana nodes are polling for the same tasks. | High | Low | Tasks are idempotent, so executing them multiple times will not result in logical error, but will degrade performance. To test for this case we add plenty of unit tests around this logic and document manual testing procedure. |
| Code should gracefully handle cases when feature X or plugin Y are disabled. | Medium | High | Unit tests will verify that any feature flag or plugin combination still results in our service operational. |
| [See more potential risk examples](https://github.com/elastic/kibana/blob/master/RISK_MATRIX.mdx) |


### For maintainers
